File: aws_knowledge_base/kb_retrieve_and_generate.js
import { fileURLToPath } from "url";
import { BedrockAgentRuntimeClient, RetrieveAndGenerateCommand } from "@aws-sdk/client-bedrock-agent-runtime";

const REGION = process.env.AWS_DEFAULT_REGION || "us-east-1";
const KNOWLEDGE_BASE_ID = process.env.KNOWLEDGE_BASE_ID || "AEC7EUSCHX";
const MODEL_ID = process.env.FOUNDATION_MODEL_ID || "meta.llama3-8b-instruct-v1:0";

const client = new BedrockAgentRuntimeClient({ region: REGION });

export const retrieveAndGenerate = async (query) => {
    const command = new RetrieveAndGenerateCommand({
        input: {
            'text': query
        },
        retrieveAndGenerateConfiguration: {
            'type': 'KNOWLEDGE_BASE',
            'knowledgeBaseConfiguration': {
                'knowledgeBaseId': KNOWLEDGE_BASE_ID,
                'modelArn': `arn:aws:bedrock:${REGION}::foundation-model/${MODEL_ID}`
            }
        }
    });
    const response = await client.send(command);

    return response;
};

// Invoke function if this file was run directly.
if (process.argv[1] === fileURLToPath(import.meta.url)) {
    const query = process.argv[2];
    if (query === undefined || query === null) {
        console.log('error: query parameter is required.');
        process.exit(1);
    }
    const response = await retrieveAndGenerate(process.argv[2]);
    console.log(response['output']['text']);
}

File: aws_knowledge_base/kb_retrieve_and_generate.py
import os
import boto3
import argparse
import sys

aws_profile = os.environ.get("AWS_PROFILE") 
region = os.environ.get("AWS_DEFAULT_REGION", "us-east-1") 
knowledge_base_id = os.environ.get("KNOWLEDGE_BASE_ID", "AEC7EUSCHX")
model_id = os.environ.get("FOUNDATION_MODEL_ID", "meta.llama3-8b-instruct-v1:0") 


def retrieve_and_generate(client, query, knowledge_base_id, model_id, region):
    return client.retrieve_and_generate(
        input={
            'text': query
        },
        retrieveAndGenerateConfiguration={
            'type': 'KNOWLEDGE_BASE',
            'knowledgeBaseConfiguration': {
                'knowledgeBaseId': knowledge_base_id,
                'modelArn': f'arn:aws:bedrock:{region}::foundation-model/{model_id}'
            }
        }
    )

def main(argv=sys.argv[1:]):
    parser = argparse.ArgumentParser()
    parser.add_argument('-q', '--query', required=True, help='The question to be asked to the LLM agent.')
    args = parser.parse_args(argv) 

    session = boto3.Session(profile_name=aws_profile)
    bedrock_agent_runtime_client = session.client("bedrock-agent-runtime", region_name=region)
    response = retrieve_and_generate(bedrock_agent_runtime_client, args.query, knowledge_base_id, model_id, region)
    return response['output']['text']

if __name__ == "__main__":
    print(main())